Lib/logging/__init__.py

```diff
@@ -1543,9 +1557,17 @@ def isEnabledFor(self, level):
         """
         Is this logger enabled for level 'level'?
         """
-        if self.manager.disable >= level:
-            return False
-        return level >= self.getEffectiveLevel()
+        try:
+            return self._cache[level]
+        except KeyError:
+            _acquireLock()
+            if self.manager.disable >= level:
+                self._cache[level] = False
+            else:
+                self._cache[level] = level >= self.getEffectiveLevel()
+            _releaseLock()
+
+            return self._cache[level]
```

Comment:
There's a still a race condition here if _clear_cache() is called just after _releaseLock(). You will need to store the value in a local variable, e.g:
```python
_acquireLock()
is_enabled = self._cache[level] = ...
_releaseLock()
return is_enabled
```